ipfs-8-to-9/migration/migration.go

```diff
@@ -0,0 +1,110 @@
+// package mg8 contains the code to perform 8-9 repository migration in
+// go-ipfs. This performs a switch to raw multihashes for all keys in the
+// go-ipfs datastore (https://github.com/ipfs/go-ipfs/issues/6815).
+package mg8
+
+import (
+	"errors"
+	"fmt"
+
+	migrate "github.com/ipfs/fs-repo-migrations/go-migrate"
+	lock "github.com/ipfs/fs-repo-migrations/ipfs-1-to-2/repolock"
+	"github.com/ipfs/go-datastore/namespace"
+
+	mfsr "github.com/ipfs/fs-repo-migrations/mfsr"
+	log "github.com/ipfs/fs-repo-migrations/stump"
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-ipfs/plugin/loader"
+	fsrepo "github.com/ipfs/go-ipfs/repo/fsrepo"
+)
+
+// Migration implements the migration described above.
+type Migration struct{}
+
+// Versions returns the current version string for this migration.
+func (m Migration) Versions() string {
+	return "8-to-9"
+}
+
+// Reversible returns false. This migration cannot be reverted, as we do not
+// know which raw hashes were actually CIDv1s. However, things should work all
+// the same as they will be treated as CIDv0s in old versions anyways.
+func (m Migration) Reversible() bool {
+	return false
+}
+
+// Apply runs the migration.
+func (m Migration) Apply(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("applying %s repo migration", m.Versions())
+
+	log.VLog("locking repo at %q", opts.Path)
+	lk, err := lock.Lock2(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+
+	log.VLog("  - verifying version is '8'")
+	if err := repo.CheckVersion("8"); err != nil {
+		return err
+	}
+
+	log.VLog("  - loading repo configurations")
+	plugins, err := loader.NewPluginLoader(opts.Path)
+	if err != nil {
+		return fmt.Errorf("error loading plugins: %s", err)
+	}
+
+	if err := plugins.Initialize(); err != nil {
+		return fmt.Errorf("error initializing plugins: %s", err)
+	}
+
+	if err := plugins.Inject(); err != nil {
+		return fmt.Errorf("error injecting plugins: %s", err)
+	}
+
+	cfg, err := fsrepo.ConfigAt(opts.Path)
+	if err != nil {
+		return err
+	}
+
+	dsc, err := fsrepo.AnyDatastoreConfig(cfg.Datastore.Spec)
+	if err != nil {
+		return err
+	}
+
+	dstore, err := dsc.Create(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer dstore.Close()
```

Comment:
Opted to go this way as I am not allowed to use Open() because ipfs repo version is lower than expected. If we depend on a newer version, might use `Open()` directly. But I'm thinking I'll still need to load the plugins.

Unfortunately did not find a pretty way to parse configuration manually and open datastores by hand. The fact that specs are `map[string]interface{}` with very different possibilities inside does not help.

Reply:
Well, I assume we'd depend on the _current_ version of go-ipfs (old repo).  Also note, we can just set `fsrepo.RepoVersion`. Take a look at `Revert` in 6-to-7.

However, what you're doing here definitely works and is probably better (no measure datastore).

Reply:
> no measure datastore

Sure? Isn't the Spec wrapping in measure ds anyways (for flatfs at least)?

Reply:
Ah, interesting. We're measuring multiple times. One for "global" datastore stats, one per datastore type. We should reconsider that (https://github.com/ipfs/go-ipfs/issues/6877).

(not all that important for the current change)